### PR TITLE
Port update_buffer_defaults

### PR DESCRIPTION
**Pull request by shaleh**
Bring per_buffer_idx, set_per_buffer_value, per_buffer_value_p along too.

### CODE REVIEW

File: rust_src/src/buffers.rs [hunk 1]
```diff
@@ -340,6 +348,14 @@ impl LispBufferRef {
             None
         }
     }
+
+    pub fn set_value(&mut self, offset: usize, value: LispObject) {
```
Comment (db48x) [inline rust_src/src/buffers.rs]: This function doesn't check that the offset is inside the buffer, so it should be `unsafe`.

File: rust_src/src/buffers.rs [hunk 2]
```diff
@@ -1006,6 +1022,12 @@ pub fn erase_buffer() -> LispObject {
     Qnil
 }
 
+pub fn per_buffer_idx(offset: isize) -> isize {
```
Comment (db48x) [inline rust_src/src/buffers.rs]: ditto

File: rust_src/src/chartable.rs [hunk 3]
```diff
@@ -186,7 +184,7 @@ impl LispSubCharTableRef {
         let mut val = self._get(idx);
 
         if is_uniprop && uniprop_compressed_form_p(val) {
-            val = unsafe { uniprop_table_uncompress(self.as_lisp_obj(), idx as libc::c_uint) };
+            val = unsafe { uniprop_table_uncompress(self.as_lisp_obj(), idx as i32) };
```
Comment (db48x) [inline rust_src/src/chartable.rs]: This change is unrelated.

File: rust_src/src/data.rs [hunk 4]
```diff
@@ -597,6 +598,29 @@ pub unsafe extern "C" fn store_symval_forwarding(
...
 }
 
+unsafe fn update_buffer_defaults(objvar: *const LispObject, newval: LispObject) {
+    // If this variable is a default for something stored
+    // in the buffer itself, such as default-fill-column,
+    // find the buffers that don't have local values for it
+    // and update them.
+    let defaults = LispBufferRef::new(&mut buffer_defaults as *mut Lisp_Buffer);
+    let defaults_as_object = defaults.as_ptr() as *const LispObject;
+    if objvar > defaults_as_object && objvar < defaults_as_object.add(1) {
```
Comment (db48x) [inline rust_src/src/data.rs]: In the original, this test is true when objvar points inside the default_buffer object, which is a long struct containing many LispObject variables.

This version, however, tests to see if objvar points between the first and second LispObject of the default_buffer.

File: rust_src/src/data.rs [hunk 5]
```diff
@@ -597,6 +598,29 @@ pub unsafe extern "C" fn store_symval_forwarding(
...
+        let offset = (objvar as *const c_char).offset_from(defaults.as_ptr() as *const c_char);
+        let idx = per_buffer_idx(offset);
```
Comment (db48x) [inline rust_src/src/data.rs]: Here we compute an offset using a `*c_char`, and then pass it to `per_buffer_idx` which adds it to a pointer of type `*LispBuffer`. That'll give the wrong result, because offsets are measured relative to the size of the item pointed to.

File: src/lisp.h [hunk 6]
```diff
@@ -3763,6 +3763,7 @@ extern void map_char_table_for_charset (void (*c_function) (Lisp_Object, Lisp_Ob
 					Lisp_Object, struct charset *,
 					unsigned, unsigned);
 extern Lisp_Object uniprop_table (Lisp_Object);
+extern Lisp_Object uniprop_table_uncompress (Lisp_Object, int);
```
Comment (db48x) [inline src/lisp.h]: more unrelated changes

Comment (shaleh) [reply src/lisp.h]: Yeah, this PR ended up with bits from another PR. I will clean it.

File: rust_src/src/data.rs (hunk 5 above)
Comment (shaleh) [reply rust_src/src/data.rs]: Good catch. I will give this code a closer look.

File: rust_src/src/data.rs (hunk 4 above)
Comment (shaleh) [reply rust_src/src/data.rs]: Thanks. I have not worked with raw pointers much lately.

File: rust_src/src/buffers.rs [hunk 7]
```diff
@@ -1020,10 +1018,10 @@ pub fn erase_buffer() {
     }
 }
 
-pub fn per_buffer_idx(offset: isize) -> isize {
-    let flags = LispBufferRef::new(unsafe { &mut buffer_local_flags as *mut Lisp_Buffer });
-    let flags_object = flags.as_ptr() as *const LispObject;
-    unsafe { (*(flags_object.add(offset as usize))).as_fixnum_or_error() as isize }
+pub unsafe fn per_buffer_idx(offset: isize) -> isize {
+    let flags = &mut buffer_local_flags as *mut Lisp_Buffer as *mut libc::c_char;
+    let obj = flags.add(offset as usize) as *const LispObject;
```
Comment (db48x) [inline rust_src/src/buffers.rs]: Since `offset` is an isize, you should probably just use the `offset` method rather than the `add` method.

File: rust_src/src/chartable.rs [hunk 8]
```diff
@@ -184,7 +186,7 @@ impl LispSubCharTableRef {
         let mut val = self._get(idx);
 
         if is_uniprop && uniprop_compressed_form_p(val) {
-            val = unsafe { uniprop_table_uncompress(self.as_lisp_obj(), idx as i32) };
+            val = unsafe { uniprop_table_uncompress(self.as_lisp_obj(), idx as libc::c_uint) };
```
Comment (db48x) [inline rust_src/src/chartable.rs]: This is also unrelated.

File: rust_src/src/data.rs (hunk 5 above)
Comment (db48x) [reply rust_src/src/data.rs]: I should probably have been clearer about this; I want the offset to still be the count of LispObjects, not the count of bytes. It reduces the amount of casting involved, and it will make it easier to change this in the future.

File: rust_src/src/remacs_sys.rs [hunk 9]
```diff
@@ -81,6 +81,7 @@ extern "C" {
         depth: libc::c_int,
         ht: Lisp_Object,
     ) -> bool;
+    pub fn uniprop_table_uncompress(table: Lisp_Object, idx: u32) -> Lisp_Object;
```
Comment (db48x) [inline rust_src/src/remacs_sys.rs]: All the `uniprop_table_uncompress` stuff is unrelated.

File: rust_src/src/buffers.rs (hunk 7 above)
Comment (shaleh) [reply rust_src/src/buffers.rs]: ok.

File: rust_src/src/data.rs (hunk 5 above)
Comment (shaleh) [reply rust_src/src/data.rs]: Ah, I was not 100% sure we could trust it. I will give that a go.

File: rust_src/src/remacs_sys.rs (hunk 9 above)
Comment (shaleh) [reply rust_src/src/remacs_sys.rs]: There should not be any uniprop changes in this PR now. I reverted all of it.

Comment (shaleh) [reply rust_src/src/remacs_sys.rs]: No worries!

**Comment by shaleh** [comment]
Thanks for the guidance @db48x 

I have been out of pointer land too long apparently :-)
